### ISSUE TITLE
CdiCoreShutdown() indiscriminately cancels application layer socket calls

### ISSUE THREAD

**Issue by yanniel** [aws/aws-cdi-sdk#35]
Calling **CdiCoreShutdown()** cancels not only the CDI SDK socket calls, but also all the sockets calls at the application layer that are not related to the CDI SDK.

**CdiCoreShutdown()**--> **SdkShutdownInternal()** --> **CdiOsShutdown()** --> **WSACleanup()**

"When WSACleanup is called, any pending blocking or asynchronous Windows Sockets calls issued by any thread in this process are canceled....". https://docs.microsoft.com/en-us/windows/win32/api/winsock/nf-winsock-wsacleanup

This pulls the rug out from under the application's feet. 

To workaround this consumers of the CDI SDK are forced to have a model where the life cycle of the application matches the life cycle of the CDI SDK usage. This seems rather restrictive. 

It would be beneficial to have a model where applications could engage/disengage the CDI SDK. With the current design “adapter resources” (ex. TX adapter buffers used to transmit payloads) are allocated during adapter initialization, but can only be de-allocated when calling **CdiCoreShutdown()**. This forces an app to keep resources around that are not being used any more during the life of the application.

Consider somewhat related case: https://github.com/aws/aws-cdi-sdk/issues/36.

OS: Windows Server 2019 Datacenter 64-bit (10.0 Build 17763)

**Comment by mhhen** [comment]
Thank you for providing additional details here beyond https://github.com/aws/aws-cdi-sdk/issues/35. We reviewed the information at: https://docs.microsoft.com/en-us/windows/win32/api/winsock/nf-winsock-wsacleanup

The resources are only freed as described below:

"There must be a call to WSACleanup for each successful call to WSAStartup. Only the final WSACleanup function call performs the actual cleanup. The preceding calls simply decrement an internal reference count in the WS2_32.DLL"

So, when **CdiCoreShutdown()** is called it will call WSACleanup once, if WSAStartup was used within the CDI-SDK. If the application has also called WSAStartup, the call to **CdiCoreShutdown()** will not actually free the Windows Socket service resources. In this case, the application has control over the resources and must call WSACleanup when it is done. We did test this condition.

Since **CdiCoreShutdown()** can be used without forcibly freeing Windows Socket service resources, does this also address your other concern about adapter resource allocation/freeing?

**Comment by yanniel** [comment]
Thanks @mhhen . This was the first issue I encountered when integrating the SDK, and it seems to have been addressed in https://github.com/aws/aws-cdi-sdk/issues/30: "Unbalanced call to WSACleanup in CdiCoreShutdown". Apologies for not realizing this issue I opened was a duplicate of a closed one.

"does this also address your other concern about adapter resource allocation/freeing?" --> not fully. **CdiCoreShutdown()** frees the resources across all adapters; but I would rather have an API call to free resources that are **adapter specific**. I am trying to do my design with the expectation that more than one adapter could be used per application (say for redundancy). [Edit]: @mhhen . I added extra details about this concern on https://github.com/aws/aws-cdi-sdk/issues/36#issuecomment-775350531

Thanks.
